### PR TITLE
Feature/add uk sanctions list

### PR DESCRIPTION
**Pull request by diginicho**
- Adds data from the uk sanctions list

### CODE REVIEW

File: cmd/server/search_handlers.go [hunk 1]
```diff
@@ -338,16 +341,38 @@ var (
...
+	result := make([]searchGather, totalLen)
+
+	var i int
+
+	for _, s := range slices {
+		i += copy(result[i:], s)
+	}
+
+	return result
+}
```
Comment (adamdecaf) [inline cmd/server/search_handlers.go:374]: I'm confused why this is needed. We should be able to append without a helper func. `append(arr, items...)`

File: pkg/csl/download_uk.go [hunk 2]
```diff
@@ -15,19 +15,36 @@ import (
...
 var (
 	// taken from https://www.gov.uk/government/publications/financial-sanctions-consolidated-list-of-targets/consolidated-list-of-targets#contents
-	publicUKDownloadURL = "https://ofsistorage.blob.core.windows.net/publishlive/2022format/ConList.csv"
-	ukDownloadURL       = strx.Or(os.Getenv("UK_CSL_DOWNLOAD_URL"), publicUKDownloadURL)
+	publicUKCSLDownloadURL = "https://ofsistorage.blob.core.windows.net/publishlive/2022format/ConList.csv"
+	ukCSLDownloadURL       = strx.Or(os.Getenv("UK_CSL_DOWNLOAD_URL"), publicUKCSLDownloadURL)
+
+	// https://www.gov.uk/government/publications/the-uk-sanctions-list
+	publicUKSanctionsListURL = "https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/1121113/UK_Sanctions_List.ods"
+	ukSanctionsListURL       = strx.Or(os.Getenv("UK_SANCTIONS_LIST_URL"), publicUKSanctionsListURL)
```
Comment (adamdecaf) [inline pkg/csl/download_uk.go:23]: We'll need to add this to the README.md

File: pkg/csl/reader_uk.go [hunk 3]
```diff
@@ -1,31 +1,37 @@
...
+	"bytes"
 	"encoding/csv"
 	"errors"
 	"fmt"
 	"io"
 	"os"
 	"strconv"
 	"strings"
+
+	"github.com/knieriem/odf/ods"
```
Comment (adamdecaf) [inline pkg/csl/reader_uk.go:13]: Q: Is there another package that can read open document files? I'm a bit worried to import this code as-is.

File: cmd/server/search_handlers.go (hunk 1 above)
Comment (diginicho) [reply cmd/server/search_handlers.go:374]: This helper appends multiple slices together. In the case of the searchGathers when a new list is added and a searchGather for it, there will need to be more variables to add each of them. For instance,
```
baseGather = append(cslGatherers, someOtherGatherers...)
euGather = []searchGather
ukGather = []searchGather
allGather = append(baseGather, euGather...)
allAllGather = append(allGather, ukGather...)
```
the intermediate lists take up space and aren't as intuitive as just having a helper functions that allocates a list of the total size of the sub lists and appends them all.

File: pkg/csl/reader_uk.go (hunk 3 above)
Comment (diginicho) [reply pkg/csl/reader_uk.go:13]: The list of libraries that read open document files is pretty slim, but I can take a look at another one.

Comment (adamdecaf) [reply pkg/csl/reader_uk.go:13]: I read the code a bit more and this library is pretty small. It's fine for now, but I am going to try and fuzz it a little.

**Comment by adamdecaf** [comment]
```
Process completed with exit code 143.
```

This seems to be Actions killing the job, but I'm not sure why. I can't find much except that the job is probably exceeding a cpu/memory/network quota of Actions.

**Comment by diginicho** [comment]
There's a 10 minute timeout by default on go tests and I am able to get this to reproduce locally. If I comment out the call to the ods library and rerun all the tests pass, so it seems to be isolated there. The thing I'm wondering now is is this a false positive? Are we really ever going to be running into a scenario in which we download these sheets in parallel?

**Comment by adamdecaf** [comment]
> There's a 10 minute timeout by default on go tests and I am able to get this to reproduce locally. If I comment out the call to the ods library and rerun all the tests pass, so it seems to be isolated there. 

I think we need to skip those tests on Actions CI then. At least for now until we can get them running/understood better. (Assuming the code is working.)

> Are we really ever going to be running into a scenario in which we download these sheets in parallel?

On startup and refresh the URLs are downloaded, but not in parallel. How many sheets are we talking about?

**Comment by diginicho** [comment]
> How many sheets are we talking about?
We are up to 9 spreadsheets at this point, some with multiple sheets/tabs.

> (Assuming the code is working.)
That's the thing the code is working just fine, so I'm not positive this test is catching anything of note.

**Comment by adamdecaf** [comment]
If we wanted to download the sheets into a directory and parse/search over them that'll help validate most of the process on each commit. We don't need to download them over CI if it's brittle.

**Comment by ckelly-digicert** [comment]
@adamdecaf After discussing this with @diginicho, I believe the best course of action at this time is to remove this particular test from the Actions CI. Perhaps we comment out the test code, with comments that explain why the test(s) are commented out for now. Since the CI is breaking due to attempting to perform a race condition check on functionality that is ultimately prone to race conditions, but for which we will never get into that scenario (only operating on one file at a time), I believe we can address this later in a different way. Thoughts?

**Comment by adamdecaf** [comment]
Can we `t.Skip("...")` the failing tests for now? That can contain a description of the problem.

**Comment by ckelly-digicert** [comment]
Absolutely
@diginicho will make the change and push it up once he's done with lunch.

**Comment by diginicho** [comment]
@adamdecaf The tests affected by the race test were skipped. In addition, I was able to trace back the source of the make test-integration error to code from the admin package:
```
// NewServer returns an admin Server instance that handles Prometheus metrics
// and pprof requests.
// Callers can use ':0' to bind onto a random port and call BindAddr() for the address.
func NewServer(addr string) *Server {
	timeout, _ := time.ParseDuration("45s")

	var listener net.Listener
	if addr == ":0" {
		listener, _ = net.Listen("tcp", "127.0.0.1:0")
	} else {
		listener, _ = net.Listen("tcp", addr)
	}
```
In this case the 45 seconds is not long enough any more to download all the spreadsheets. I would recommend increasing that to 90s to be sure.

**Comment by adamdecaf** [comment]
I think we should preload those 9 new documents so they don't need to be downloaded in the test.

**Comment by adamdecaf** [comment]
I created https://github.com/moov-io/base/pull/268 to allow configurable timeouts on the admin server.

File: makefile [hunk 4]
```diff
@@ -119,11 +119,13 @@ cover-web:
...
+# for now this shouldn't hold up out CI pipeline
 test-integration: clean-integration
-	docker-compose up -d
-	sleep 30
-	curl -v http://localhost:9094/data/refresh # hangs until download and parsing completes
-	./bin/batchsearch -local -threshold 0.95
+	# docker compose up -d
+	# sleep 75
+	# time curl -v --max-time 90 http://localhost:9094/data/refresh # hangs until download and parsing completes
+	# ./bin/batchsearch -local -threshold 0.95
```
Comment (adamdecaf) [inline makefile:128]: I was thinking we comment this out in the Actions workflow, not in the makefile.

**Comment by adamdecaf** [comment]
Can I get a copy of the race detector errors? I want to dig into what the problem is there.

**Comment by adamdecaf** [comment]
```
$ gotest ./pkg/csl/... -count 1 -race -short -run TestReadUKSanctionsList -v 
=== RUN   TestReadUKSanctionsList
--- PASS: TestReadUKSanctionsList (267.09s)
PASS
ok  	github.com/moov-io/watchman/pkg/csl	267.960s
```

Parsing a 14MB file takes 4.5mins? That seems pretty inefficient so I'm curious what that library is doing..

**Comment by adamdecaf** [comment]
It looks like the majority of the time is spent doing xml decoding. We could look at the SAX-based parsing (e.g. using .Token()) for some performance gains. Spending multiple minutes is going to cause folks unexpected problems when upgrading. 

Screenshot 2022-12-13 at 8 56 22 AM
Screenshot 2022-12-13 at 8 56 31 AM

[uk_csl_cpu.prof.txt](https://github.com/moov-io/watchman/files/10219137/uk_csl_cpu.prof.txt)
[uk_csl_mem.prof.txt](https://github.com/moov-io/watchman/files/10219138/uk_csl_mem.prof.txt)

**Comment by adamdecaf** [comment]
From Github support, 

```
I'm seeing that the linked run you shared appears to be the last instance of receiving exit code 143. Accompanying that result, an error annotation for that same job indicates:

The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
Like you had mentioned, this usually occurs when processes on the Actions runner are starved of resources which can cause the possibility of a graceful shutdown if communication between the Actions service and runner becomes idle.

Were any changes made on your end to potentially lessen the resource consumption by your workflow operations?

Altogether, it looks like this issue has not been a consistent blocker recently. If this error would reoccur, we encourage evaluating the point in the workflow where this shutdown is happening - and seeing if there are any changes that can be made to spread resource utilization across a longer runtime to not stall the VM.
```

**Pull request merged**